scripts/trim_indel_ends.py

```diff
@@ -0,0 +1,452 @@
+
+    # return prefix_indels, suffix_indels[::-1]
+    return get_len_type(prefix_indels), get_len_type(suffix_indels[::-1])
+
+def trim_seq(prefix_type, prefix_len, suffix_type, suffix_len, seq, num, threshold=10) -> str:
+    """Trim input sequence according to suffix and prefix number of indels
+
+    Args:
+        prefix_type (str): type of edits at the beginning of the cigar, ("I" or "D")
+        prefix_len (int): length of prefix edits
+        suffix_type (str):type of edits at the end of the cigar, ("I" or "D")
+        suffix_len (int): length of suffix edits
+        seq (str): input sequence to be trimmed
+        num (int): The order of the sequence in alignment. If a sequence is the first in alignment, deletion means that a character is removed. 
+        Otherwise, insertion means trimming.
+        threshold (int): maximum number of bases trimmed
+
+    Returns:
+        str: trimmed sequence
+
+    Example:
+        Input1: trim_seq("D", 1, "I",2, "ATCGAAG", 1)
+        Output1: "TCGAAG"  (The first character is trimmed but the suffix is not trimmed)
+        
+        Input2: trim_seq("D", 1, "I",2, "ATCGAAG", 2)
+        Output2: "ATCGA"  (The suffix of length 2 is trimmed but the prefix is not trimmed)
+
+    """
+    assert num == 1 or num == 2, "trim_seq(): unknown ranking of input sequence!"
+
+    if num == 2:
+        if prefix_type == "D":
+            seq = seq[min(threshold, prefix_len):]
+        if suffix_type == "D":
+            seq = seq[:-min(threshold, suffix_len)]
+    if num == 1:
+        if prefix_type == "I":
+            seq = seq[min(threshold, prefix_len):]
+        if suffix_type == "I":
+            seq = seq[:-min(threshold, suffix_len)]
+    return seq
+
+def trim_from_cigar(cigar_str, sequence1, sequence2, threshold=10, mode="lenient"):
+    """Given two sequences and the cigar string describing the alignment between them, trim the start and end, if the 
+    alignment starts and/or ends with indels.
+
+    Args:
+        cigar_str (str): cigar_string
+        sequence1(str): aligned sequence 1 to be trimmed
+        sequence2(str): aligned sequence 2 to be trimmed
+        threshold(int): maximum number of bases trimmed
+        mode(str): strict or lenient
+            - strict: only trim if the trimmed result will be all matches
+            - lenient: trim whenever there are indels at the edge of the strings
+    Returns:
+        str: trimmed sequence 1
+        str: trimmed sequence 2
+    """
+
+    (prefix_type, prefix_length), (suffix_type, suffix_length) = get_prefix_and_suffix_indels(cigar_str, mode)
+
+    if prefix_length == 0 and suffix_length == 0:
+        return sequence1, sequence2
+
+    global total_trim
+    total_trim += 1
+
+    sequence1 = trim_seq(prefix_type, prefix_length, suffix_type, suffix_length, sequence1, 1, threshold)
+    sequence2 = trim_seq(prefix_type, prefix_length, suffix_type, suffix_length, sequence2, 2, threshold)
+
+    global effective_trim
```

Comment:
Generally not a big fan of global variables -- could we think about making a class object to track this?

Reply:
global variables are hard to test

Reply:
I don't particularly like global variables either, especially in a group project like this. 

Let's try to remember these and make sure there's no weird interactions. 

At some point, these probably should be rewritten to not use global variables. CC @ame9yu

Reply:
Comments resolved by enclosing the two variables with a trim tracker class in the most current commit